### ISSUE TITLE
NER predictions from pretrained models

### ISSUE THREAD

**Issue by yrahul3910** [dmis-lab/biobert#50]
I'm trying to run NER on a dataset using a pretrained model. It wouldn't let me run this without either `--do_eval=true` or `--do_train=true`, and I used the advice in [this comment](https://github.com/dmis-lab/biobert/issues/39#issuecomment-519350834) to fix it. However, it appears that predicting depends on a `label2id.pkl`, that can only be generated when either of the two flags above are true. My next attempt was to copy a `devel.tsv` from one of the preprocessed datasets to my data directory and set `--do_eval=true`, but I wonder if there's a better way to do this.

Further, I get this output at the end:  

```
INFO:tensorflow:***** Running prediction*****
I1023 03:59:30.193573 140324722628480 run_ner.py:595] ***** Running prediction*****
INFO:tensorflow:  Num examples = 0
I1023 03:59:30.193796 140324722628480 run_ner.py:596]   Num examples = 0
INFO:tensorflow:  Batch size = 8
I1023 03:59:30.193949 140324722628480 run_ner.py:597]   Batch size = 8
Traceback (most recent call last):
  File "run_ner.py", line 646, in <module>
    tf.app.run()
  File "/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/usr/local/lib/python3.6/dist-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "run_ner.py", line 610, in main
    with open(token_path, 'r') as reader:
FileNotFoundError: [Errno 2] No such file or directory: './tmp/token_test.txt'
```

So I have two questions:  
(a) My `test.tsv` is just rows of sentences: is this correct?  
(b) I'm not sure how I'm supposed to get the `token_test.txt` file to exist?

**Comment by wonjininfo** [comment]
Hi, 
One simple way to bypass `raise ValueError("At least one of `do_train` or `do_eval` must be True.")` error is to set 
(1) `--do_train=true` 
(2) `--num_train_epochs=0.1` (or less than pre-trained ckpt)
(3) `--output_dir=(as your pre-trained model)` 
(4) (optional) `--init_checkpoint=(as your pretrained model)`
and this will skip training steps. 

Since our code (and also BERT) checks whether the model in **the output folder** is already trained, setting  `--num_train_epochs` less than the pre-trained model in the output folder will make the code skip training and just perform prediction. 

For example,
``` 
python run_ner.py \
    --vocab_file=$BIOBERT_DIR/vocab.txt \
    --bert_config_file=$BIOBERT_DIR/bert_config.json \
    --init_checkpoint=$BIOBERT_DIR/biobert_model.ckpt \
    --data_dir=$NER_DIR/ \
    --do_train=true \
    --do_predict=true \
    --num_train_epochs=0.1 \
    --output_dir=/tmp/bioner/(pre-trained dir) 
```

I will check with other colleagues about the [line](https://github.com/dmis-lab/biobert/blob/7a3c96e3fca2129fe74d78f9b707397fedd4cbd9/run_ner.py#L477),
Please use our "bypass method" until then. 
Thanks!
Wonjin

**Comment by yrahul3910** [comment]
Thanks for the help, Wonjin. It seems to have gone further now, but I get this output at the end:

```
INFO:tensorflow:Saving checkpoints for 27 into tmp/model.ckpt.
I1031 20:04:58.358532 140402530543424 basic_session_run_hooks.py:606] Saving checkpoints for 27 into tmp/model.ckpt.
INFO:tensorflow:Loss for final step: 74.07268.
I1031 20:05:01.774962 140402530543424 estimator.py:368] Loss for final step: 74.07268.
INFO:tensorflow:training_loop marked as finished
I1031 20:05:01.775710 140402530543424 error_handling.py:96] training_loop marked as finished
INFO:tensorflow:***** Running prediction*****
I1031 20:05:01.932822 140402530543424 run_ner.py:595] ***** Running prediction*****
INFO:tensorflow:  Num examples = 0
I1031 20:05:01.932993 140402530543424 run_ner.py:596]   Num examples = 0
INFO:tensorflow:  Batch size = 8
I1031 20:05:01.933125 140402530543424 run_ner.py:597]   Batch size = 8
Traceback (most recent call last):
  File "run_ner.py", line 646, in <module>
    tf.app.run()
  File "/home/ryedida/.local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/home/ryedida/.local/lib/python3.6/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/home/ryedida/.local/lib/python3.6/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "run_ner.py", line 610, in main
    with open(token_path, 'r') as reader:
FileNotFoundError: [Errno 2] No such file or directory: 'tmp/token_test.txt'
```
Where `tmp/` is my `--output_dir` argument. My guess then is that I should create this file using the tokenizers. Is that correct?

**Comment by soumyavhasure** [comment]
I tried using your bypass method and I am still not able to get the `token_test.txt` and `label_test.txt` files to be created in the output directory. Also, I don't have a `(pre-trained dir)` in the output directory. Please advise.

**Comment by wonjininfo** [comment]
Hi  Yedida @yrahul3910  ,
It looks like you didn't fine tuned NER model at all (from the line `INFO:tensorflow:Saving checkpoints for 27 into tmp/model.ckpt.`). 
You need to fine-tune our weights on NER first and use the method I introduced to you. 
BioBERT weights, which we share, doesn't contain pre-trained model for NER task. (If we skip some details, we can say that we are sharing some kind of word embeddings.)

So, you need to finetune first, and use that "pretrained" weight for your next experiments. (`token_test.txt` will be generated while you finetune our weights )

Also, hi @soumyavhasure , it looks like you are having same problem. Would you please check this comment and try again?

**Comment by yrahul3910** [comment]
Hi Wonjin,

By fine-tune, do you mean using this command:  

```
python run_ner.py \
    --do_train=true \
    --do_eval=true \
    --vocab_file=$BIOBERT_DIR/vocab.txt \
    --bert_config_file=$BIOBERT_DIR/bert_config.json \
    --init_checkpoint=$BIOBERT_DIR/biobert_model.ckpt \
    --num_train_epochs=10.0 \
    --data_dir=$NER_DIR/ \
    --output_dir=/tmp/bioner/
```
I ran this, but towards the end, it fails:  

```
INFO:tensorflow:Starting evaluation at 2019-11-04T01:54:36Z
I1104 01:54:36.791576 140363460466496 evaluation.py:255] Starting evaluation at 2019-11-04T01:54:36Z
INFO:tensorflow:Graph was finalized.
I1104 01:54:37.244819 140363460466496 monitored_session.py:240] Graph was finalized.
2019-11-04 01:54:37.296876: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2019-11-04 01:54:37.297791: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: GRID T4-16Q major: 7 minor: 5 memoryClockRate(GHz): 1.59
pciBusID: 0000:02:00.0
2019-11-04 01:54:37.298055: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcudart.so.10.0'; dlerror: libcudart.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298151: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcublas.so.10.0'; dlerror: libcublas.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298237: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcufft.so.10.0'; dlerror: libcufft.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298330: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcurand.so.10.0'; dlerror: libcurand.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298416: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcusolver.so.10.0'; dlerror: libcusolver.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298506: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcusparse.so.10.0'; dlerror: libcusparse.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298589: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcudnn.so.7'; dlerror: libcudnn.so.7: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298609: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1663] Cannot dlopen some GPU libraries. Skipping registering GPU devices...
2019-11-04 01:54:37.298632: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-11-04 01:54:37.298648: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2019-11-04 01:54:37.298658: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
WARNING:tensorflow:From /home/ryedida/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py:1276: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
W1104 01:54:37.298877 140363460466496 deprecation.py:323] From /home/ryedida/.local/lib/python3.6/site-packages/tensorflow/python/training/saver.py:1276: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.
Instructions for updating:
Use standard file APIs to check for files with this prefix.
INFO:tensorflow:Restoring parameters from ./tmp/model.ckpt-2725
I1104 01:54:37.300084 140363460466496 saver.py:1280] Restoring parameters from ./tmp/model.ckpt-2725
INFO:tensorflow:Running local_init_op.
I1104 01:54:38.046426 140363460466496 session_manager.py:500] Running local_init_op.
INFO:tensorflow:Done running local_init_op.
I1104 01:54:38.108790 140363460466496 session_manager.py:502] Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2019-11-04-01:57:45
I1104 01:57:45.648785 140363460466496 evaluation.py:275] Finished evaluation at 2019-11-04-01:57:45
INFO:tensorflow:Saving dict for global step 2725: eval_f = 0.015847528, eval_precision = 0.27336955, eval_recall = 0.008681207, global_step = 2725, loss = 170.35191
I1104 01:57:45.649134 140363460466496 estimator.py:2039] Saving dict for global step 2725: eval_f = 0.015847528, eval_precision = 0.27336955, eval_recall = 0.008681207, global_step = 2725, loss = 170.35191
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 2725: ./tmp/model.ckpt-2725
I1104 01:57:46.387177 140363460466496 estimator.py:2099] Saving 'checkpoint_path' summary for global step 2725: ./tmp/model.ckpt-2725
INFO:tensorflow:evaluation_loop marked as finished
I1104 01:57:46.387874 140363460466496 error_handling.py:96] evaluation_loop marked as finished
INFO:tensorflow:***** Eval results *****
I1104 01:57:46.388306 140363460466496 run_ner.py:577] ***** Eval results *****
INFO:tensorflow:  eval_f = 0.015847528
I1104 01:57:46.388427 140363460466496 run_ner.py:579]   eval_f = 0.015847528
INFO:tensorflow:  eval_precision = 0.27336955
I1104 01:57:46.388524 140363460466496 run_ner.py:579]   eval_precision = 0.27336955
INFO:tensorflow:  eval_recall = 0.008681207
I1104 01:57:46.388617 140363460466496 run_ner.py:579]   eval_recall = 0.008681207
INFO:tensorflow:  global_step = 2725
I1104 01:57:46.388702 140363460466496 run_ner.py:579]   global_step = 2725
INFO:tensorflow:  loss = 170.35191
I1104 01:57:46.388782 140363460466496 run_ner.py:579]   loss = 170.35191
INFO:tensorflow:***** Running prediction*****
I1104 01:57:46.591777 140363460466496 run_ner.py:595] ***** Running prediction*****
INFO:tensorflow:  Num examples = 0
I1104 01:57:46.591975 140363460466496 run_ner.py:596]   Num examples = 0
INFO:tensorflow:  Batch size = 8
I1104 01:57:46.592140 140363460466496 run_ner.py:597]   Batch size = 8
Traceback (most recent call last):
  File "run_ner.py", line 646, in <module>
    tf.app.run()
  File "/home/ryedida/.local/lib/python3.6/site-packages/tensorflow/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/home/ryedida/.local/lib/python3.6/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/home/ryedida/.local/lib/python3.6/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "run_ner.py", line 610, in main
    with open(token_path, 'r') as reader:
FileNotFoundError: [Errno 2] No such file or directory: './tmp/token_test.txt'
```
Running the prediction code after this fails again, with the same complaint.

**Comment by wonjininfo** [comment]
Hi @yrahul3910 
That is strange. I will take a look as soon as I go back to my office. Thanks for sharing details!

**Comment by wonjininfo** [comment]
Hi @yrahul3910 
I executed the same command line you mentioned. It works without any problem. 
Just in case, have you changed any of the source code? 
I am investigating this now. 
Would you tell me which commit are you on? Are you on the latest commit?

**Comment by wonjininfo** [comment]
-----
Btw, I don't think this is a cause of your error but, your hardware configurations might have some error. 

From your output looks like some cuda-related liblaries are missing. :
```2019-11-04 01:54:37.298055: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcudart.so.10.0'; dlerror: libcudart.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298151: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcublas.so.10.0'; dlerror: libcublas.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298237: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcufft.so.10.0'; dlerror: libcufft.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298330: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcurand.so.10.0'; dlerror: libcurand.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298416: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcusolver.so.10.0'; dlerror: libcusolver.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298506: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcusparse.so.10.0'; dlerror: libcusparse.so.10.0: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298589: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Could not dlopen library 'libcudnn.so.7'; dlerror: libcudnn.so.7: cannot open shared object file: No such file or directory
2019-11-04 01:54:37.298609: W tensorflow/core/common_runtime/gpu/gpu_device.cc:1663] Cannot dlopen some GPU libraries. Skipping registering GPU devices...
2019-11-04 01:54:37.
```

Normally, it looks like :
```
INFO:tensorflow:Graph was finalized.
2019-11-05 11:30:31.904289: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 0
2019-11-05 11:30:31.904365: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-11-05 11:30:31.904379: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977]      0
2019-11-05 11:30:31.904389: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 0:   N
2019-11-05 11:30:32.162928: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 23019 MB memory) -> physical GPU (device: 0, name: Tesla P40, pci bus id: 0000:04:00.0, compute capability: 6.1)
INFO:tensorflow:Restoring parameters from /tmp/ner/model.ckpt-68
```

**Comment by yrahul3910** [comment]
It appears I was about one commit behind, but running `git pull` only updated the readme...so my code at least is up to date.

I agree that the CUDA/CUDNN libraries missing isn't the cause of the issue either. I'll get those installed. I suspect the issue may have something to do with lines 586-587 that try removing the `token_path` variable (which seems to be the `token_test.txt` file. Do you think commenting those out might help? Thanks for being so patient with this, I really appreciate it!

**Comment by wonjininfo** [comment]
Hi @yrahul3910 ,
Would you double-check that `os.remove(token_path)` precede `filed_based_convert_examples_to_features(predict_examples, label_list` code? 

Function `convert_single_example` inside function `filed_based_convert_examples_to_features` has `wirte_tokens`. 
In normal case, this function writes `token_test.txt`. (Thanks Chan Ho! @parao0628) 

So just commenting out `os.remove(token_path)` won't help to solve the fundamental problem. There must be some problems regarding `filed_based_convert_examples_to_features` and your environment.

**Comment by yrahul3910** [comment]
I feel this may be the issue: 

```
INFO:tensorflow:***** Running prediction*****
I1104 01:57:46.591777 140363460466496 run_ner.py:595] ***** Running prediction*****
INFO:tensorflow:  Num examples = 0
I1104 01:57:46.591975 140363460466496 run_ner.py:596]   Num examples = 0
INFO:tensorflow:  Batch size = 8
```

So it would appear that it can't find test examples in my `test.tsv` file. It's likely that this file is formatted wrongly. Ideally, I should just have single words (or sentences) in each line, and the prediction mode should...predict? In my `NER_DIR` right now is `train_dev.tsv`, `devel.tsv` (these two I simply copied from the `s800` directory), and `test.tsv` (which I suspect is incorrectly formatted).

**Comment by wonjininfo** [comment]
Hi @yrahul3910 , 
It seems like its a different issue. 
Would you please check the format of `test.tsv`? 
Please open another issue if you still have the same problem.

To the best of my knowledge, we checked the dataset multiple times.

**Comment by demongolem** [comment]
I see it is closed, but I do have a note on one of the key questions asked by the OP.  The file token_test.txt was generated for me when I did the fine tuning, but then on a later run when I was attempting to use the model just to predict it, the file disappeared.  I know this because my eclipse window showed the file, then when I refreshed it disappeared.

In my case, I had a backup on another server, so I could restore this crucial file.  But why would token_test.txt or how could token_test.txt disappear on subsequent non-training runs?
